### PR TITLE
Add the debugging docs for the NuGet products

### LINKED ISSUE #8947: Port the dev guide to GitHub
Right now it's a One Note. 
It's more useful in the repo in some sort of a doc. 

Ideas from NuGet/NuGet.Client#3180

- NuGet and target frameworks

- NuGet + dependencies, how to add a new dependency in the NuGet client. Think source-build.

- All NuGet to NuGet dependencies are via a ProjectReference (implied via all NuGet assemblies version the same :) )

- How to add a new assembly section.

- We need a section for NuGet Error codes

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8947
Fixes: https://github.com/NuGet/Client.Engineering/issues/137
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
Port many of the internal debugging docs to the repo.
Note that after this is merged I will delete the sanity guide. 

The content is pretty much the same. 

I have also added the debugging helper scripts. Those are not perfect, but mostly just work. 
Similarly, unless you see a functional bug, I'd appreciate if any changes to it happen in a future PR. 

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Docs
Validation:

### CODE REVIEW

File: docs/debugging.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+NuGet functions as an extension on top of Visual Studio.
+However NuGet is also considered a system component which means it cannot be managed in the same way as other Visual Studio extensions. However, it is possible to install a locally built NuGet in a Visual Studio instance.
+
+#### Installing the NuGet extension in Visual Studio
+
+The build generates a vsix artifact in the `artifacts/VS15/` folder. From the [Developer Command Prompt for VS](https://docs.microsoft.com/en-us/dotnet/framework/tools/developer-command-prompt-for-vs) run the `VSIXInstaller.exe` with the vsix path as the first argument. Alternatively VS configures the default action for VSIX files.
+
+#### Uninstalling the NuGet extension from Visual Studio
+
+Given that NuGet is a system component, you cannot use the extensions manager in Visual Studio. Go back to your [Developer Command Prompt for VS](https://docs.microsoft.com/en-us/dotnet/framework/tools/developer-command-prompt-for-vs) and run `VSIXInstaller.exe /d:NuGet.72c5d240-f742-48d4-a0f1-7016671e405b`.
```
Comment (donnie-msft) [inline docs/debugging.md]: Need the keyword "downgrade" in here somehow...

```suggestion
Given that NuGet is a system component, you cannot use the extensions manager in Visual Studio to downgrade your NuGet extension to its original version. Go back to your [Developer Command Prompt for VS](https://docs.microsoft.com/en-us/dotnet/framework/tools/developer-command-prompt-for-vs) and run `VSIXInstaller.exe /d:NuGet.72c5d240-f742-48d4-a0f1-7016671e405b`.
```

File: docs/debugging.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+Testing the NuGet Visual Studio functionality is equally as easy as testing NuGet.exe.
+The start-up project is [NuGet.VisualStudio.Client](../src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj). Starting this will launch the [experimental instance](https://docs.microsoft.com/en-us/visualstudio/extensibility/the-experimental-instance) of Visual Studio.
+
+### Testing the build of NuGet in Visual Studio
+
+NuGet functions as an extension on top of Visual Studio.
+However NuGet is also considered a system component which means it cannot be managed in the same way as other Visual Studio extensions. However, it is possible to install a locally built NuGet in a Visual Studio instance.
+
+#### Installing the NuGet extension in Visual Studio
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
#### Installing a custom version of the NuGet extension in Visual Studio
```

File: docs/debugging.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+The build generates a vsix artifact in the `artifacts/VS15/` folder. From the [Developer Command Prompt for VS](https://docs.microsoft.com/en-us/dotnet/framework/tools/developer-command-prompt-for-vs) run the `VSIXInstaller.exe` with the vsix path as the first argument. Alternatively VS configures the default action for VSIX files.
+
+#### Uninstalling the NuGet extension from Visual Studio
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
#### Uninstalling a custom version of the NuGet extension from Visual Studio
```

Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
#### Downgrading the NuGet extension from Visual Studio
```

File: docs/debugging.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+Given that NuGet is a system component, you cannot use the extensions manager in Visual Studio. Go back to your [Developer Command Prompt for VS](https://docs.microsoft.com/en-us/dotnet/framework/tools/developer-command-prompt-for-vs) and run `VSIXInstaller.exe /d:NuGet.72c5d240-f742-48d4-a0f1-7016671e405b`.
+
+#### NuGet in Visual Studio assembly location
+
+Each Visual Studio instance has it's own root install directory. Relative to the root directory, the NuGet assemblies can be are in `Common7/IDE/CommonExtensions/Microsoft/NuGet`.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
Each Visual Studio instance has its own root install directory. Relative to the root directory, the NuGet assemblies can be found in `Common7/IDE/CommonExtensions/Microsoft/NuGet`.
```

File: docs/debugging.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+### Code pointers for NuGet in Visual Studio
+
+Visual Studio extensibility has the concept of VSPackages. This allows extensions to add capabilities to the UI and all of the flows within the IDE. The NuGet client itself ships 2 packages, which can be considered the entry points for the whole component, You wouldn’t normally debug the initialization of these packages of course, but they author the whole NuGet in VS experience in one or another.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
Visual Studio extensibility has the concept of VSPackages. This allows extensions to add capabilities to the UI and all of the flows within the IDE. The NuGet client itself ships 2 packages, which can be considered the entry points for the whole component. You wouldn’t normally debug the initialization of these packages of course, but they author the whole NuGet in VS experience in one or another.
```

File: docs/debugging.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+The 2 packages in question are:
+
+* [NuGetPackage](../src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs) - enables most of the UI NuGet functionality in Visual Studio.
+* [RestoreManagerPackage](../src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs) - enables the restore functionality in Visual Studio.
+
+The NuGet in Visual Studio experience is scattered across many assemblies as can be seen from the [project overview](project-overview.md).
+Some of the other points are summarized below:
+
+* Restore operations - [SolutionRestoreWorker](../src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs) - All Visual Studio restores start here. Currently only a solution restore is possible. Depending on the project package management style there are different code paths to look into.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
* Restore operations - [SolutionRestoreWorker](../src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs) - All Visual Studio restores start here. Currently, only a solution restore is possible. Depending on the project package management style, there are different code paths to look into.
```

File: docs/debugging.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+* Package Manager UI operations - All menu items are defined the [NuGetTools.vsct](../src/NuGet.Clients/NuGet.Tools/NuGetTools.vsct). The main control for the Package Manager UI is [PackageManagerControl](../src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs).
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
* Package Manager UI operations - VS context menu integration is defined in [NuGetTools.vsct](../src/NuGet.Clients/NuGet.Tools/NuGetTools.vsct). The main control for the Package Manager UI is [PackageManagerControl](../src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs).
```

File: docs/debugging.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+* Package Manager Console operations - The powershell cmdlets are defined in the [NuGet.PackageManagement.Cmdlets](../src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/NuGet.PackageManagement.PowerShellCmdlets.csproj) project, more specifically in the [cmdlets](../src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/Cmdlets) folder.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
* Package Manager Console operations - The PowerShell cmdlets are defined in the [NuGet.PackageManagement.Cmdlets](../src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/NuGet.PackageManagement.PowerShellCmdlets.csproj) project, more specifically in the [cmdlets](../src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/Cmdlets) folder.
```

File: docs/debugging.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+Alternatively, the NuGet.targets are just wired into the build in a certain way and that can be replicated without a complete installation of the extension through the powershell helper functions as shown in the [PowerShell helper scripts](../scripts/nuget-debug-helpers.ps1). Examples can be found in the linked scripts.
+
+If you are testing a debug build, to debug just set the environment variable defined in [RestoreTask](../src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs), currently `DEBUG_RESTORE_TASK`.
+
+### Testing and debugging pack in MSBuild
+
+The naive approach here is to install the package to the project and just run `msbuild -t:pack` on it.
+
+Alternatively, if you are testing on SDK based projects, the NuGet.Build.Tasks.Pack.targets are just wired into the build in a certain way and that can be replicated without a complete installation of the extension through the powershell helper functions as shown in the [PowerShell helper scripts](../scripts/nuget-debug-helpers.ps1). Examples can be found in the linked scripts.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
Alternatively, if you are testing on SDK based projects, the NuGet.Build.Tasks.Pack.targets are just wired into the build in a certain way and that can be replicated without a complete installation of the extension through the PowerShell helper functions as shown in the [PowerShell helper scripts](../scripts/nuget-debug-helpers.ps1). Examples can be found in the linked scripts.
```

File: docs/debugging.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+dotnet.exe restore works the exact same way `msbuild -t:restore` works. The one difference is that .NET Core based MSBuild cannot build the same project types as .NET Framework MSBuild can.
+Most of the tips in [Testing and debugging restore in MSBuild](testing-and-debugging-restore-in-msbuild) still apply.
+
+If you want to test dotnet.exe explicitly, refer to [Patching dotnet.exe to test the NuGet functionality](patching-dotnet-exe-to-test-the-nuget-functionality).
+
+### Debugging pack task in dotnet.exe
+
+dotnet.exe pack works the exact same way `msbuild -t:pack` works. The added benefit is that it is always available in all SDK based projects, which is what dotnet.exe supports.
+Most of tips in [Testing and debugging pack in MSBuild](testing-and-debugging-pack-in-msbuild) still apply.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
Most of the tips in [Testing and debugging pack in MSBuild](testing-and-debugging-pack-in-msbuild) still apply.
```

File: docs/debugging.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+Functionality such as `dotnet.exe add package` or `dotnet list package`, is implemented in [NuGet.CommandLine.XPlat](../src\NuGet.Core\NuGet.CommandLine.XPlat\NuGet.CommandLine.XPlat.csproj).
+There are 2 ways to debug this:
+
+* Given that [NuGet.CommandLine.XPlat](../src\NuGet.Core\NuGet.CommandLine.XPlat\NuGet.CommandLine.XPlat.csproj) is an exe, you can set it as the startup project and run it as you would any other command line project in Visual Studio. Note that some commands list arguments in a different order to the dotnet cli.
+
+* Patch the SDK by referring to [Patching dotnet.exe to test the NuGet functionality](patching-dotnet-exe-to-test-the-nuget-functionality).
+
+After you have patched it, refer to the environment variable in the [NuGet.CommandLine.XPlat](../src\NuGet.Core\NuGet.CommandLine.XPlat\NuGet.CommandLine.XPlat.csproj) entry point.
+If you are testing a debug build, to debug just set the environment variable defined in [Program.cs](../src\NuGet.Core\NuGet.CommandLine.XPlat\NuGet.CommandLine.XPlat.csproj), currentl DEBUG_NUGET_XPLAT.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
If you are testing a debug build, to debug just set the environment variable defined in [Program.cs](../src\NuGet.Core\NuGet.CommandLine.XPlat\NuGet.CommandLine.XPlat.csproj), currently, `DEBUG_NUGET_XPLAT`.
```

File: docs/debugging.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+### Patching dotnet.exe to test the NuGet functionality
+
+There are 2 ways to patch dotnet.exe with the latest NuGet bits.
+
+* Refer to the [PowerShell helper scripts](../scripts/nuget-debug-helpers.ps1) for a helper script to patch a zip of the SDK that can acquired from the [dotnet installer](https://github.com/dotnet/installer/blob/master/README.md#installers-and-binaries). Note that when running you [might](https://github.com/dotnet/runtime/blob/master/docs/project/dogfooding.md) need to disable MULTI_LEVEL_LOOKUP. dotnet.exe has a special logic for the discovering the SDK, and it's possible it's discovering a different SDK from the one you patched.
```
Comment (donnie-msft) [inline docs/debugging.md]:
```suggestion
* Refer to the [PowerShell helper scripts](../scripts/nuget-debug-helpers.ps1) for a helper script to patch a zip of the SDK that can be acquired from the [dotnet installer](https://github.com/dotnet/installer/blob/master/README.md#installers-and-binaries). Note that when running you [might](https://github.com/dotnet/runtime/blob/master/docs/project/dogfooding.md) need to disable MULTI_LEVEL_LOOKUP. dotnet.exe has a special logic for the discovering the SDK, and it's possible it's discovering a different SDK from the one you patched.
```

**Review by donnie-msft — APPROVED** [review]
Didn't dive into the script as much as the guide. Think this guide will be very useful when working on unfamiliar areas (or forgetting) and needing to debug.

**Comment by nkolev92** [comment]
> Didn't dive into the script as much as the guide. 

@donnie-msft 

Yeah, the scripts work just fine for now, so not worth the effort. 

The first 3 are well vetted, I'm fairly confident cause I use those a lot. 
It's the add-NuGetToCLI one that could use some love, but there's not much to earn from that work.

File: docs/debugging.md (hunk 2 above)
Comment (nkolev92) [reply docs/debugging.md]: This requires some link changes. I'll make sure they are addressed

**Pull request merged**